src/journal/journald-stream.c

```diff
@@ -470,20 +470,46 @@ static int stdout_stream_scan(StdoutStream *s, bool force_flush) {
                 p += skip;
         }
 
-        if (force_flush && remaining > 0) {
-                p[remaining] = 0;
+        if (p > s->buffer) {
+                memmove(s->buffer, p, remaining);
+                s->length = remaining;
+        }
+
+        return 0;
+}
+
+static int stdout_stream_flush(StdoutStream *s, size_t new_rcv_len) {
+        char *p;
+        size_t current_len;
+        int r;
+
+        assert(s);
+
+        p = s->buffer;
+        current_len = s->length;
+
+        /* This function flush the current content of the stream buffer. This part was
+         * already scanned by stdout_stream_scan(), so no need to do it again.
+         * Then the new received length content is moved to the beginning of the buffer,
+         * and the buffer length is updated with this new length */
+
+        if (current_len > 0) {
+                char saved_char;
+
+                saved_char = p[current_len];
```

Comment:
why save the char? it's the full length of the buffer we force out, and hence what comes after it doesn't matter, it's old data, possibly even uninitialized...

We always keep space for one more byte in the buffer so that we can NUL terminate everything...

Reply:
OK imagine the following sequence:
- Received from a PID 123 the following output on stdout: `abcde` without a new line
- Received from a PID 124 the following output on stdout: `12345` without a new line
- ...

So:
- The string `abcde` is going to be in the buffer, it will be scanned, since no new line, the scan function does nothing and leave `abcde` in the buffer.
- Then the PID 124 output `12345`, so the buffer is going to contain `abcde12345`
- The code detects that the PID did change, so we need to flush `abcde` first, then scan `12345`
- Do do that we need to be a  `\0` where is the `1` in the buffer in order to send it using `stdout_stream_line()`